### ISSUE TITLE
[feature] component shouldnt require consumers to also use components.

### ISSUE THREAD

**Issue by fulara** [conan-io/conan#7496]
I have prepared a repository that you can use to validate the flow without spending your own time on preparing example:
https://github.com/fulara/testing-components/tree/snapshot-github

note this is pretty much a duplicate of: https://github.com/conan-io/conan/issues/7475
However its a question there, and I am requesting the feature here  ( actually I am more leaning towards raising a bug, but its not mentioned anywhere in the docs, so that would be kinda unfair of me . ) 

Short explanation:
liba -> lowest library - wants to expose components.
libb -> library that depends on liba.
bin -> binary that depends on libb
libb-component - equivalent of libb but using dummy component
bin-component - equivalent of bin but depending on libb-component

I would like to have components in liba. 
only a single layer above liba the libb  in my example knows about liba.

So in this case i define `liba`  package_info as follows:
```
    def package_info(self):
        self.cpp_info.components["publisher"].libs = ["publisher"]
        self.cpp_info.components["requester"].libs = ["requester"]
```

and in the libb package_info I am kinda stuck, I dont know how to refer to the components of liba without having to use my component so I am stuck with package_info like this:
```
    def package_info(self):
        self.cpp_info.libs = ["libb"] # because i am using cpp_info.libs conan is causing dependence on whole liba (all libs from it. )
        #self.cpp_info.requires = ["liba::requester"] #i wish this line was working as expected.
```

Now I am going to binary, I am just using cmake_find_package in this case and this is how the cmakelist of th ebin looks like:

```
cmake_minimum_required(VERSION 2.8)
project(MyHello CXX)

find_package(libb)

add_executable(bin hello.cpp)
target_link_libraries(bin libb::libb)
```

When I build the bin now though this is the linker command i can see:
```
/usr/bin/c++    -rdynamic CMakeFiles/bin.dir/hello.cpp.o  -o bin   .... libpublisher.a ..librequester.a
```
I would like publisher.a  not to be present on that list. 

anyway, I created libb-components and bin-components, it is working as expected ( libpublisher.a  is not appearing on the list ).

I guess it would be somewhat fine if I could do it like this in the libb-component:
```
    def package_info(self):
        self.cpp_info.components[self.name].libs = ["libb"]
        self.cpp_info.components[self.name].requires = ["liba::requester"]
```
but there is an assert against it.

So in short, it looks to me like component forces usage of the component up the stack, and provide unnexpected behavior if someone does not do so, which is counter intuitive.

I think possibly the most natural way of doing this would be adding the .requires on the cpp_info that would allow dependance on individual components, and only allow depending on the components via this.
I saw some assert that told me that i forgot to add .requires to my components when depending on something, so that could be probably added also there.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
This seems a valid feature request. I agree that possibly the desired UX would be:

```python
self.cpp_info.requires = ["liba::requester"]
```

Tagging this for having a look and analyzing if doable.

**Comment by memsharded** [comment]
Yes, that is true, this seems a duplicate of https://github.com/conan-io/conan/issues/7475, lets close this one and convert the previous existing one to a feature request.
